api/src/main/java/org/pmiops/workbench/actionaudit/auditors/SumoLogicAuditorImpl.kt

```diff
@@ -0,0 +1,132 @@
+    private val workspaceService: WorkspaceService,
+    private val userDao: UserDao,
+    private val clock: Clock,
+    @Qualifier("ACTION_ID") private val actionIdProvider: Provider<String>
+) : SumoLogicAuditor {
+
+    /**
+     * Converts a DB-internal user ID to the corresponding VM name. Note: for now, we use the same
+     * logic used by ClusterController, but if that logic changes, we may want to adapt this code
+     * to support searching for VMs with both the older and newer naming scheme.
+     */
+    private fun dbUserToClusterName(dbUser: DbUser): String {
+        return ClusterController.clusterNameForUser(dbUser)
+    }
+
+    /**
+     * Converts the AoU-chosen cluster name into the actual VM name as reported by Google Cloud's
+     * flow logs. Empirically, an "-m" suffix is added to the VM name.
+     *
+     * Note: this pattern may change if the Leo team switches to using raw VMs instead of Dataproc
+     * clusters!
+     */
+    private fun dbUserToVmName(dbUser: DbUser): String {
+        return dbUserToClusterName(dbUser) + "-m"
+    }
+
+    override fun fireEgressEvent(event: EgressEvent) {
+        // Load the workspace via the GCP project name
+        val dbWorkspace = this.workspaceService.getByNamespace(event.projectName)
+        if (dbWorkspace == null) {
+            fireFailedToFindWorkspace(event)
+            return
+        }
+        // Using service-level creds, load the FireCloud workspace ACLs to find all members
+        // of the workspace. Then attempt to find the user who aligns with the named VM.
+        val userRoles = workspaceService.getFirecloudUserRoles(dbWorkspace.workspaceNamespace,
+                dbWorkspace.firecloudName)
+        val vmOwner = userRoles
+                .map { userDao.findUserByUsername(it.email) }
+                .filter { dbUserToVmName(it).equals(event.vmName) }
+                .firstOrNull()
+
+        var agentEmail = ""
+        var agentId = 0L
+        if (vmOwner != null) {
+            agentEmail = vmOwner.username
+            agentId = vmOwner.userId
+        } else {
+            // If the VM name doesn't match a user on the workspace, we'll still log an
+            // event in the target workspace, but with nulled-out user info.
+            logger.warning(String.format("Could not find a user for VM name %s in project %s", event.vmName, event.projectName))
+        }
+
+        actionAuditService.send(ActionAuditEvent(
+                timestamp = clock.millis(),
+                actionId = actionIdProvider.get(),
+                actionType = ActionType.HIGH_EGRESS,
+                agentType = AgentType.USER,
+                agentId = agentId,
+                agentEmailMaybe = agentEmail,
+                targetType = TargetType.WORKSPACE,
+                targetIdMaybe = dbWorkspace.workspaceId,
+                // TODO: Maybe what would fit best here would be a "comment" field on the AcionAuditEvent and
```

Comment:
There are a couple of ways to do this, depending on what level of abstraction you want.

If you want commenting to be an action that someone can take, then we can make an ActionType of COMMENT, and apply it to any of the target types. We could even have a target type of ACTION and use the action's ID as the TARGET_ID.

I'm hesitant to make a new column for comment, as it would mean different rows under the same Action could have different comments. That's not a problem per se, but it's difficult to enforce a consistent usage pattern when lots of folks are using the system, and querying might be icky.

Reply:
Thanks – I guess the most pragmatic solution will be to call it an ActionType and tie two rows together by a single Action ID: one for the high-egress event, and one for the comment.

I'd feel a little awkward to have to send two events for this, even though it's definitely just one event-plus-a-log-message. But I can wrap it in a helper function on the ActionAuditService, so it probably won't affect usage too much.

Reply:
PTAL at the [latest commit](https://github.com/all-of-us/workbench/pull/2992/commits/c5ccc982bdc8c4bb7b790470420fbda199e29007) – I took a stab at something like what I described above.